### PR TITLE
Support for composite types

### PR DESCRIPTION
**Pull request by dburkart**
This PR tracks adding composite support:

- [x] Scanner / Parser support
- [x] Schema support
- [x] Value support

### CODE REVIEW

File: pkg/schema/objects_test.go [hunk 1]
```diff
@@ -44,10 +44,4 @@ func TestJSONMarshal(t *testing.T) {
 	if string(b) != `"[10]int32"` {
 		t.Fail()
 	}
-
-	tc := Composite{Keys: []string{"x", "y"}, Values: []Object{Type{Name: "int32"}, Type{Name: "int32"}}}
```
Comment (gideonw) [inline pkg/schema/objects_test.go:48]: Reason for removal?

File: pkg/schema/parser_test.go [hunk 2]
```diff
@@ -113,6 +114,30 @@ func TestParseArray(t *testing.T) {
 	}
 }
 
+func slicesEqualStr(a, b []string) bool {
```
Comment (gideonw) [inline pkg/schema/parser_test.go:117]: https://pkg.go.dev/slices

This is now in the std library as of 1.21. If you don't want to maintain this and instead implement the EqualFunc

Comment (gideonw) [reply pkg/schema/parser_test.go:117]: DIdn't notice this was a test file lol

**Review by gideonw — APPROVED** [review]
Looks great man. I didn't see anything that stood out to me as an error or bug.

File: pkg/schema/objects_test.go (hunk 1 above)
Comment (dburkart) [reply pkg/schema/objects_test.go:48]: We don't use the JSON marshaling stuff anywhere for composites (I'm actually not sure we use it for any schema object type). And it didn't work at all lol.

The JSON marshaler was just erroring out.

File: pkg/query/plan/function.go [hunk 3]
```diff
@@ -57,12 +58,20 @@ func (f *Function) Visit(node ast.ASTNode) ast.Visitor {
...
 			f.results[n] = types.BinaryOp(f.results[n.Left], n.Op, f.results[n.Right])
-		case *ast.TupleElementNode:
+		case *ast.ElementNode:
 			result, ok := f.symbols[n.Identifier.Value()]
 			if !ok {
 				panic(fmt.Sprintf("Symbol %s did not resolve!", n.Identifier.Value()))
 			}
-			f.results[n] = types.TupleVal(result)[types.IntVal(n.Subscript.Val)]
+
+			switch s := n.Subscript.(type) {
```
[the remark this answers is not in the document]
Comment (dburkart) [reply pkg/query/plan/function.go:67]: I think they're different IIRC, we're just lax the other way around: in places where we expect a string, we now also accept an identifier (sometimes).

Particularly for composite keys, we'll accept both quoted and unquoted (eg "key" and key). Mostly just for ergonomics.

**Pull request merged**